mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+
+// processStatement extracts schema information from MYSQL
+// statements, updating Conv with new schema information, and returning
+// true if INSERT statement is encountered.
+func processStatement(conv *internal.Conv, stmt ast.StmtNode) bool {
+
+	switch ddlStmt := stmt.(type) {
+	case *ast.CreateTableStmt:
+		if conv.SchemaMode() {
+			processCreateTable(conv, ddlStmt)
+		}
+	case *ast.AlterTableStmt:
+		if conv.SchemaMode() {
+			processAlterTable(conv, ddlStmt)
+		}
+	case *ast.SetStmt:
+		if conv.SchemaMode() {
+			processSetStmt(conv, ddlStmt)
+		}
+	case *ast.InsertStmt:
+		return true
+	default:
+		conv.SkipStatement(reflect.TypeOf(stmt).String())
+	}
+	return false
+}
+
+func processSetStmt(conv *internal.Conv, stmt *ast.SetStmt) {
+	if stmt.Variables != nil && len(stmt.Variables) > 0 {
+		for _, variable := range stmt.Variables {
+			if variable.Name == "TIME_ZONE" {
+				value := variable.Value
+				switch val := value.(type) {
+				case *driver.ValueExpr:
+					if val.GetValue() == nil {
+						logStmtError(conv, stmt, fmt.Errorf("found nil value in set statement"))
+						return
+					}
+					conv.TimezoneOffset = fmt.Sprintf("%v", val.GetValue())
+				default:
+					logStmtError(conv, stmt, fmt.Errorf("found %s type value in set statement", reflect.TypeOf(val)))
+					return
+				}
+			}
+		}
+	}
+}
+
+func processCreateTable(conv *internal.Conv, stmt *ast.CreateTableStmt) {
+	var colNames []string
+	colDef := make(map[string]schema.Column)
+	if stmt.Table == nil {
+		logStmtError(conv, stmt, fmt.Errorf("table is nil"))
+		return
+	}
+	tableName, err := getTableName(stmt.Table)
+	if err != nil {
+		logStmtError(conv, stmt, fmt.Errorf("can't get table name: %w", err))
+		return
+	}
+	columns := stmt.Cols
+	var keys []schema.Key
+	for _, element := range columns {
+		colname, col, isPk, err := processColumn(conv, tableName, element)
+		if err != nil {
+			logStmtError(conv, stmt, err)
+			return
+		}
+		colNames = append(colNames, colname)
+		colDef[colname] = col
+		if isPk {
+			keys = append(keys, schema.Key{Column: colname})
+		}
+	}
+	conv.SchemaStatement(reflect.TypeOf(stmt).String())
+	conv.SrcSchema[tableName] = schema.Table{
+		Name:        tableName,
+		ColNames:    colNames,
+		ColDefs:     colDef,
+		PrimaryKeys: keys}
+	for _, constraint := range stmt.Constraints {
+		processConstraint(conv, tableName, constraint, "CREATE TABLE")
+	}
+}
+
+func processConstraint(conv *internal.Conv, table string, constraint *ast.Constraint, s string) {
+	st := conv.SrcSchema[table]
+	switch ct := constraint.Tp; ct {
+	case ast.ConstraintPrimaryKey:
+		checkEmpty(conv, st.PrimaryKeys, s) // Drop any previous primary keys.
+		st.PrimaryKeys = toSchemaKeys(constraint.Keys)
+		// In Spanner, primary key columns are usually annotated with NOT NULL,
+		// but this can be omitted to allow NULL values in key columns.
+		// In MySQL, the primary key constraint is a combination of
+		// NOT NULL and UNIQUE i.e. primary keys must be NOT NULL.
```

Comment:
Should we also set the unique bit?

I know we don't currently do this for postgres, but we probably should.

Reply:
https://github.com/cloudspannerecosystem/harbourbridge/blob/79a6ee343f846cf1623f75311d9925322478bf1e/mysql/mysqldump.go#L337-L346

Do you mean to say that for `case ast.ConstraintPrimaryKey` we should add `cd.Unique = true` also along with `cd.NotNull = true`?

Reply:
Yes sorry I should have been more precise.

Adding this will not change the spanner schema we generate.
But it does change the source schema.
And I think it is more consistent -- what's your take?

Reply:
Yes true. I had not done that just to be in sync with `pgdump.go` but I also think it would be a good practice. Updated!
https://github.com/cloudspannerecosystem/harbourbridge/blob/99db369b2b826fc24ffbb760f3b3700cd0e04b0f/mysql/mysqldump.go#L222-L241